protocol/verification/persistence.go

```diff
@@ -0,0 +1,252 @@
+}
+
+func (p *Persistence) GetVerificationRequests() ([]Request, error) {
+	rows, err := p.db.Query("SELECT from_user, to_user, challenge, response, requested_at, verification_status, replied_at FROM verification_requests")
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	var result []Request
+	for rows.Next() {
+		var vr Request
+		var requestedAt int64
+		var repliedAt int64
+		err = rows.Scan(&vr.From, &vr.To, &vr.Challenge, &vr.Response, &requestedAt, &repliedAt, &vr.RequestStatus)
+		if err != nil {
+			return nil, err
+		}
+
+		vr.RequestedAt = time.Unix(requestedAt, 0)
+		vr.RepliedAt = time.Unix(repliedAt, 0)
+		result = append(result, vr)
+	}
+	return result, nil
+}
+
+func (p *Persistence) GetVerificationRequestFrom(contactID string) (*Request, error) {
+	var vr Request
+	var requestedAt int64
+	var repliedAt int64
+
+	err := p.db.QueryRow(`SELECT from_user, to_user, challenge, response, requested_at, verification_status, replied_at FROM verification_requests WHERE from_user = ?`, contactID).Scan(
+		&vr.From,
+		&vr.To,
+		&vr.Challenge,
+		&vr.Response,
+		&requestedAt,
+		&vr.RequestStatus,
+		&repliedAt,
+	)
+
+	switch err {
+	case sql.ErrNoRows:
+		return nil, nil
+	case nil:
+		vr.RequestedAt = time.Unix(requestedAt, 0)
+		vr.RepliedAt = time.Unix(repliedAt, 0)
+		return &vr, nil
+	default:
+		return nil, err
+	}
+}
+
+func (p *Persistence) GetVerificationRequestSentTo(contactID string) (*Request, error) {
+	var vr Request
+	var requestedAt int64
+	var repliedAt int64
+
+	err := p.db.QueryRow(`SELECT from_user, to_user, challenge, response, requested_at, verification_status, replied_at FROM verification_requests WHERE to_user = ?`, contactID).Scan(
+		&vr.From,
+		&vr.To,
+		&vr.Challenge,
+		&vr.Response,
+		&requestedAt,
+		&vr.RequestStatus,
+		&repliedAt,
+	)
+
+	switch err {
+	case sql.ErrNoRows:
+		return nil, nil
+	case nil:
+		vr.RequestedAt = time.Unix(requestedAt, 0)
+		vr.RepliedAt = time.Unix(repliedAt, 0)
+		return &vr, nil
+	default:
+		return nil, err
+	}
+}
+
+func (p *Persistence) SaveVerificationRequest(vr *Request) error {
+	if vr == nil {
+		return errors.New("invalid verification request provided")
+	}
+	_, err := p.db.Exec(`INSERT INTO verification_requests (from_user, to_user, challenge, response, requested_at, verification_status, replied_at) VALUES (?, ?, ?, ?, ?, ?, ?)`, vr.From, vr.To, vr.Challenge, vr.Response, vr.RequestedAt.Unix(), vr.RequestStatus, vr.RepliedAt.Unix())
+	return err
+}
+
+func (p *Persistence) AcceptContactVerificationRequest(contactID string, response string) error {
+	result, err := p.db.Exec("UPDATE verification_requests SET response = ?, replied_at = ?, verification_status = ? WHERE from_user = ?", response, time.Now().Unix(), RequestStatusACCEPTED, contactID)
+	if err != nil {
+		return err
+	}
+
+	numRows, err := result.RowsAffected()
+	if err != nil {
+		return err
+	}
+
+	if numRows == 0 {
+		return errors.New("verification request not found")
```

Comment:
Would you consider something like? If you've got multiple instances of the same error being returned, consuming code can check for a specific err via direct comparison `if err == ErrVerificationRequestNotFound { do something }`, also it saves you time if you need to call the same error in other places.

Obviously the `var()` block would go at the top in the package context.

```suggestion
var (
	ErrVerificationRequestNotFound = errors.New("verification request not found")
)

		return ErrVerificationRequestNotFound 
```